### ISSUE TITLE
Player count in the community server browser

### ISSUE THREAD

**Issue by yourmnbbn** [yourmnbbn/smext-fakequeries#4]
In the 'Internet' tab of the community server browser, player count value displayed is not from the query.

**Comment by toorisrael** [comment]
I've done some tests in CSS, as it still uses symbols and 'internet' tab is working the same way as in CSGO. However still with no success, probably still doing something wrong with detour?
https://pastebin.com/WWcKPWmm

**Comment by yourmnbbn** [comment]
Yeah, I've done that test either, it's a wrong way. Not updating user data won't stop player count being displayed. This is not how master server count players. But I've done a test calling `ISteamGameServer::SendUserDisconnect` to fake connect myself while I'm still in server, the player count changed from 1 to 0. When an user trying to connect, the server request the master server to help to authenticate user. When they left, server notify the master server. And this is how the master server count real players count, I suppose. So tyring to hack this number I guess is actually to simulate connection process just like how `CSteam3Server::NotifyClientConnect` does. But I have no idea how will it work now, more tests should be done.

**Comment by toorisrael** [comment]
Does Valve already know what's going on here? xD
![image](https://user-images.githubusercontent.com/4301144/145690943-edef32de-28f8-4704-a305-a2c7fff97821.png)
Any ideas how does it work now?
EDIT: something with BeginAuthSession/EndAuthSession now I guess. And probably not possible to bypass because done by client game?
EDIT: Oh, it looks like fakes are  always working in internet tab now without any modifications (or it's just a bug?). Looks like it's still bug for some time after first running the game, showing fakes as real.

**Comment by yourmnbbn** [comment]
It has been BeginAuthSession/EndAuthSession in csgo for a long time. SendUserConnectAndAuthenticate it's not used in csgo, as for SendUserDisconnect you can use it to test something but it's not how the normal player disconnect from the server, it's for players who disconnect from hltv server. So it's not surprising they deprecate these functions.

I haven't been working on this recently, I don't see this as a flaw or bug because it's not in the query level, but challenge is. Since all the major problems about query has been solved, I suppose. I'll take a look at this later when I'm available.

**Comment by toorisrael** [comment]
OK, at least today I've learned much more about detours and getting signatures. So what I've tried:
detouring CSteam3Server::NotifyClientDisconnect to return - changes nothing in player count, still real player count
detouring CSteam3Server::NotifyClientConnect to return false - will cause steam validation reject (just a test)
detouring CSteam3Server::SendUpdatedServerDetails to return - changes nothing in player count, still real player count
detouring CBaseClient::PerformDisconnection to return - causes error spam and detour loop when client timed out after some time, but still it changes nothing in player count
detouring CBaseClient::Disconnect to return - detour loop when client timed out after some time, changes nothing in player count
To me it looks like player is constantly communicating with Steam, instantly after disconnecting getting deauth on the server (event if csgo.exe crashed):
![1](https://user-images.githubusercontent.com/4301144/145716080-bbb1244b-9c55-4a6f-a29e-b264496eadfe.png)
![2](https://user-images.githubusercontent.com/4301144/145716083-b64c6f0c-3432-4df7-a6ae-f657115de5b6.png)
If you have any advices what I can test, let me know.

**Comment by yourmnbbn** [comment]
I don't think the engine library is the key, steamclient library is, just like how I fix the challenge. If you look at how they build A2S_PLAYER response, you will find this(in windows binary, on linux it's just offsets changed)
```asm
mov     edx, [edi+23E0h]
add     edx, [edi+23CCh]
add     edx, [edi+23F8h]
```
You can easily find this function by string reference `"> 255 players in ReplyPlayers."`

EDI is the pointer of the CGameServer instance, a singleton I suppose. EDX is the player count result including fake clients. [edi+23E0h] is the human players count. [edi+23F8h] is the fake clients count. Days ago I've tested detouring nothing but only change [edi+23E0h] to 0, the player count in the browser stays 0. But changing this to a higher value does not goes as expected, in my test case whatever value I change it still shows 1 player, only 0 will change. So I guess there must be some kind of communication during steam heart beat, it will use that value to loop the connected human players and send something to master server. 

But I haven't confirmed other circumstances. Like if there are 2 human players, change this to 1 will it show 1 in the browser? 
To figure this issue out is much harder than other binaries because there are no symbols of steamclient binary ever since, as well as the leaked code. 

EDIT: All the steamapi implementations are in the steamclient library, the test I've mentioned before of `ISteamGameServer::SendUserDisconnect`, calling which I believe can changes that value either.

**Comment by toorisrael** [comment]
> because there are no symbols of steamclient binary ever since

There are some, but extremally old.
![image](https://user-images.githubusercontent.com/4301144/145722553-e8d93fed-d121-4867-a086-9029346a28c9.png)
https://file.io/g7JJ1oLN2wu8

**Comment by yourmnbbn** [comment]
Ah, I don’t know that, thanks! But the link is broken.

**Comment by toorisrael** [comment]
https://www24.zippyshare.com/v/P6PHuCKy/file.html

**Comment by toorisrael** [comment]
So yesterday I've tested detouring few functions from steamclient, which led me to nothing, but was a good lesson for me, as at first I didn't know how to hook steamclient, example in case someone is searching for this:
`void *pLibrary = dlsym(dlopen("steamclient.so", RTLD_NOW), "CreateInterface");`
Detouring ReplyPlayers, or even blocking Heartbeat does nothing to players count for me.

Today, I've tested [this](https://github.com/vanz666/noVAC) with VAC banned account and after joining server with this account, players count on master server stays at 0. This plugin is detouring `CSteam3Server::OnValidateAuthTicketResponse`, so for now I think Steam is getting real players count from this.
However, there is a bug which is happening from time to time, especially when restarting server, when server is listed on master server, but is showing player count from A2S_INFO (only in game/internet tab, I haven't tested this with Steam API yet). I don't know how to reproduce this every time, but it's possible that could be exploited.

**Comment by yourmnbbn** [comment]
I'll take a look at this later. The lastest api change broke what I have tested before, the `ISteamGameServer::SendUserDisconnect` and change CGameServer member variable no longer cause value change. So yeah, good to hear another catch.

About the bug you've mentioned, I've never encountered this bug, for me the value is always the same. I guess maybe it's because something is late initialized.

EDIT: As for getting the base pointer of module, I'm using the method taken from sourcemod. Work on both platforms.
https://github.com/yourmnbbn/smext-fakequeries/blob/a90206b654d7d168fbf1b37fba107099d225cd10/symbolhelper.h#L77-L114

**Comment by toorisrael** [comment]
![image](https://user-images.githubusercontent.com/4301144/146054770-0fb57097-0153-412e-8aca-18bb775e7dcb.png)
This time it happened to me after detouring CBaseServer::GetNumHumanPlayers, worked for about 5 minutes (all fakes on screenshot). I'm wondering if opening some library like engine.so doesn't trigger this, because it happened to me so much times after loading my test extensions.

**Comment by yourmnbbn** [comment]
That's weird, is it happening every time after reloading or just on the first time it's loaded?

**Comment by toorisrael** [comment]
Sadly no, it happens just randomly on first loading extension / restarting server after some time. I've checked Steam api last time when it happend, and it was 0/64, but in game it was 60/64 also for my friends.

I've done a test with [nosteam](https://se7en.ws/emulyator-revemu-bir3yk-poslednyaya-versiya/) CSGO client (and server with RevEmu allowing Steam+nosteam players), and nosteam players are not visible on the internet tab, which again suggests it's based on Steam auth.

There are few (well, actually a ton of) russian servers like this:
![image](https://user-images.githubusercontent.com/4301144/146191094-d0d9057e-f017-4818-a91e-bf403f41d1ec.png)
The main server is on port 27015, on another IP in this case. However those fake servers really exist, reporting fake player count to master, and redirecting players to main server. They've also got VAC disabled, which again suggests something with Steam auth. Maybe you can figure something out.

EDIT:
Also detouring CSteam3Server::NotifyClientConnect to always return true and bypass all authentications makes connected client invisible on master server (0/64).

**Comment by yourmnbbn** [comment]
Yeah, maybe it's about the authentications. Good catch, [here](https://github.com/perilouswithadollarsign/cstrike15_src/blob/f82112a2388b841d72cb62ca48ab1846dfcc11c8/public/steam/steam_gameserver.h#L17) is also a proof I guess. I'll look into this later. 

And I've heard about those fake servers before, I've tried to create fake servers like this long ago but every time failed on log on by token. It just don't call the callback that I have registered just like how the leaked code does. I must have missed something, It's interesting though, but I didn't go further investigation and have abandoned that.

**Comment by sapsanDev** [comment]
Please update the signatures.

**Comment by yourmnbbn** [comment]
> Please update the signatures.

@sapsanDev  Is there any signature lookup failures or unexpected behavior caused by signatures not unique? I don't see any updates on dedicated server today.

**Comment by sapsanDev** [comment]
@yourmnbbn Swears at "ValidateChallengeFunc"

**Comment by yourmnbbn** [comment]
@sapsanDev I've confirmed on both supported platforms that all the signatures and features are working fine now, so if you have encountered weird behavior with the extension, you can open another issue and provide more detail about it. If you updated to the latest version from an older version, make sure to update the gamedata as well.

**Comment by sapsanDev** [comment]
@yourmnbbn I compiled your assembly for csgo, but when it starts it gives out:
[12]  file "fakequeries.ext.2.csgo.so": Look up ValidateChallengeFunc signature failed

**Comment by yourmnbbn** [comment]
@sapsanDev Can you provide the version of you sourcemod and metamod on server, as well as the version of sdk you are compiling against? And can you test is it working with the binary in release attachments on your server? Thanks.

**Comment by sapsanDev** [comment]
@yourmnbbn Metamod 1.11 - build 1145, Sourcemod Version: 1.10.0.6528

**Comment by yourmnbbn** [comment]
@sapsanDev Well, can you provide the following information? I guess it’s a system-specified problem. 
- Is it working with [this binary](https://github.com/yourmnbbn/smext-fakequeries/releases/tag/fakequeries_v1.3.0)?
- OS information and version. 

Due to sourcemod officially support only `server`, `engine` and `matchmaking_ds` libraries in gameconfig, getting address from other libraries by signatures need extra work. So maybe my implementation is not working for all the versions of the platforms. I’m not sure where goes wrong so far. I need feedback from the released binary which is confirmed working on my test platforms. 

EDIT: If you don’t need fix for some third query tools you can simply use version 1.2.2. Without this fix the extension will still work. But will break on those tools which use A2S_RULES request for a challenge number and then use it to request A2S_INFO and A2S_PLAYER response.

**Comment by toorisrael** [comment]
I'm quite sure he's russian, and all russians use no steam. He will probably need to change steamclient.so to steamclient_valve.so in the extension part loading this library, or whatever name original stemclient library he has.

**Comment by yourmnbbn** [comment]
@toorisrael Thanks! I don’t know that. 

@sapsanDev As what he has mentioned, you should check if there are bin/steamclient.[so/dll] exists. If there is not, you should make some modifications to the source code. You can either ignore the steamclient funtions detour, because it’s probably invalid signatures for your steamclient implementation library. Or you can change this to the real path of your steamclient implementation library. 
https://github.com/yourmnbbn/smext-fakequeries/blob/a90206b654d7d168fbf1b37fba107099d225cd10/symbolhelper.h#L15

**Comment by sapsanDev** [comment]
@toorisrael, @yourmnbbn  fun) CSGO is free to play, no need for nosteam)
Accordingly, the original steam libraries are present in the bin folder

OS : Linux Debian 10 GNU 2.28
_Is it working with this binary?_
The problem is the same

**Comment by xMaZax** [comment]
> @toorisrael Thanks! I don’t know that.
> 
> @sapsanDev As what he has mentioned, you should check if there are bin/steamclient.[so/dll] exists. If there is not, you should make some modifications to the source code. You can either ignore the steamclient funtions detour, because it’s probably invalid signatures for your steamclient implementation library. Or you can change this to the real path of your steamclient implementation library.
> 
> https://github.com/yourmnbbn/smext-fakequeries/blob/a90206b654d7d168fbf1b37fba107099d225cd10/symbolhelper.h#L15

The problem is relevant to me as well. I use SM 1.11/Debian 10 linux

> I'm quite sure he's russian, and all russians use no steam. He will probably need to change steamclient.so to steamclient_valve.so in the extension part loading this library, or whatever name original stemclient library he has.

Not true)

**Comment by toorisrael** [comment]
And you're russian too xD Just check if your steamclient.so file size is about 32MB.
Just to clarify, if your're using some nosteam+steam modification, it will look like this:
![image](https://user-images.githubusercontent.com/4301144/146508128-b45a6da1-651f-4362-99d7-bf5b82dbbbfa.png)

**Comment by xMaZax** [comment]
> And you're russian too xD Just check if your steamclient.so file size is about 32MB. Just to clarify, if your're using some nosteam+steam modification, it will look like this: ![image](https://user-images.githubusercontent.com/4301144/146508128-b45a6da1-651f-4362-99d7-bf5b82dbbbfa.png)

I don't use no-steam, I don't think anyone uses any emulators (cs:go free game).


@yourmnbbn I think it's worth using this option to get a pointer.
https://github.com/alliedmodders/sourcemod/commit/b62f3326111e5f02892f28eb983981e

**Comment by sapsanDev** [comment]
@toorisrael No need to use an emulator) Steamclient size 32 mb)

**Comment by yourmnbbn** [comment]
> I think it's worth using this option to get a pointer.

Yeah, it is a better implementation. 

@sapsanDev , @xMaZax Can you test if commit 41bfb14 fix the signature failure?

**Comment by sapsanDev** [comment]
@yourmnbbn , Yes it worked

**Comment by yourmnbbn** [comment]
@sapsanDev Looks like my guess was right. No error log as well? This time steamclient signature failure won’t block the extension from being loaded. It just throw an error to the log.

**Comment by sapsanDev** [comment]
@yourmnbbn The log is clean.

**Comment by Astronanus** [comment]
Please @yourmnbbn can you guys compile this for linux CSS with latest SM, MM and gamedata? I would really appreciate as I am not as skilled! I tried but to no avail

**Comment by toorisrael** [comment]
It won't work in CSS without many modifications. First of all, there is no ISteamGameServer in hl2sdk-css. This can be easily fixed, but there are no functions like "NET_ReceiveRawPacket", "CSteamSocketMgr" in CSS.

**Comment by Astronanus** [comment]
> It won't work in CSS without many modifications. First of all, there is no ISteamGameServer in hl2sdk-css. This can be easily fixed, but there are no functions like "NET_ReceiveRawPacket", "CSteamSocketMgr" in CSS.

roger that.
Even if its just faking bots as real players in fav tab…Cause thats nuf for me i dont wanna do more then that. There was this cool ext by Master Xykon that would let us set tags,map and bots but its broken for sm1.10

**Comment by toorisrael** [comment]
> There was this cool ext by Master Xykon that would let us set tags,map and bots but its broken for sm1.10

Can you send me a link to this?

**Comment by stupix** [comment]
Could this solve the problem
`CUtlVector<CBaseClient*>	m_Clients;		// array of up to [maxclients] client slots.`

**Comment by yourmnbbn** [comment]
> Could this solve the problem 
`CUtlVector<CBaseClient*> m_Clients; // array of up to [maxclients] client slots.`

@stupix I don't think so. [The test here](https://github.com/yourmnbbn/smext-fakequeries/issues/4#issuecomment-994773071) indicates that this is probably more related to the steam authentication. So far the available approach seems to be either playing around with the authentication or finding a stable way to reproduce the bug which is also mentioned in that comment.

**Comment by sev4a** [comment]
@[yourmnbbn](https://github.com/yourmnbbn)
look at the way here how they bypass the server manager https://github.com/NiDE-gg/sm-ext-a2sqcache/blob/3327ddfc2991d95e6433ec58f7b4509b13dcc8e8/extension.cpp

**Comment by yourmnbbn** [comment]
@sev4a Quick looked through the code, I didn't see any bypassing. It just cache the response information and use the cached value to respond to the query. It's useful when you have large amount of concurrent queries because it avoid many unnecessary function calls. All it does are still on the query level, which is the same as this extension.

**Comment by toorisrael** [comment]
[This](https://github.com/Leystryku/leysourceengineclient) can be easily modified to connect and authenticate csgo player in console mode on VAC disabled server. However - it's not enough. There will be no communication beetween client/server in game and client will timeout in seconds, because it doesn't support protobuf. If someone would like to add protobuf support - it should work, but it will require:
1) VAC disabled on server
2) multiple Steam instances run in Windows with fake clients in console mode
3) some automatization

If you know something like this for linux, or want to add protobuf support - let me know, I will try to help.
I haven't tried to run this on pre-protobuf csgo server version, but possibly this should work too. I'm not sure if that old server will show on master (probably not, but maybe with some bypasses?).

**Comment by yourmnbbn** [comment]
> [This](https://github.com/Leystryku/leysourceengineclient) can be easily modified to connect and authenticate csgo player in console mode on VAC disabled server. However - it's not enough. There will be no communication beetween client/server in game and client will timeout in seconds, because it doesn't support protobuf. If someone would like to add protobuf support - it should work, but it will require:
> 
> 1. VAC disabled on server
> 2. multiple Steam instances run in Windows with fake clients in console mode
> 3. some automatization
> 
> If you know something like this for linux, or want to add protobuf support - let me know, I will try to help. I haven't tried to run this on pre-protobuf csgo server version, but possibly this should work too. I'm not sure if that old server will show on master (probably not, but maybe with some bypasses?).

Well I like this project. It's really useful in some cases. But because of [this](https://github.com/Leystryku/leysourceengineclient/blob/2b379f581df485f428242c557287e40089f3d0e0/src/oob.cpp#L109) it still requires a valid account to authenticate a single fake player, let alone we have to figure out how to run multiple steam with different account(Is this easy? I don't think so). BTW I get it work on both secure and insecure server, so I don't think the first condition is a must.

Apart from those, I am interested in getting it fully functioning in csgo, not as a valid fix of this issue though, I just think it's useful.

**Comment by toorisrael** [comment]
That's true it requires real accounts, I don't thinks it's possible to fake players count on master server without authenticating real accounts. However - CSGO is free to play, making multiple accounts should not be a problem. About 2 years ago I've tested running multiple CSGO instances in Sandboxie and it was working fine, however it required [VAC Bypass](https://github.com/danielkrupinski/VAC-Bypass-Loader) to play MM/Wingman.

**Comment by yourmnbbn** [comment]
> That's true it requires real accounts, I don't thinks it's possible to fake players count on master server without authenticating real accounts. However - CSGO is free to play, making multiple accounts should not be a problem. About 2 years ago I've tested running multiple CSGO instances in Sandboxie and it was working fine, however it required [VAC Bypass](https://github.com/danielkrupinski/VAC-Bypass-Loader) to play MM/Wingman.

[Here](https://github.com/yourmnbbn/tiny-csgo-client) is a completely refactored version of leysourceengineclient, for cross-platform capability. It's far from complete, I have only implemented and tested the basic connection. If you're interested in  that project you can help to complete it, I would really appreciate any help.

**Comment by HenryTownshand** [comment]
progress in the case?

**Comment by ua-player-ua** [comment]
Is it still not working?
